scrapy/core/engine.py

```diff
@@ -88,20 +93,26 @@ def __init__(
         self.signals: SignalManager = crawler.signals
         assert crawler.logformatter
         self.logformatter: LogFormatter = crawler.logformatter
-        self.slot: Slot | None = None
+        self._slot: _Slot | None = None
         self.spider: Spider | None = None
         self.running: bool = False
         self.paused: bool = False
-        self.scheduler_cls: type[BaseScheduler] = self._get_scheduler_class(
-            crawler.settings
-        )
-        downloader_cls: type[Downloader] = load_object(self.settings["DOWNLOADER"])
-        self.downloader: Downloader = downloader_cls(crawler)
-        self.scraper: Scraper = Scraper(crawler)
         self._spider_closed_callback: Callable[[Spider], Deferred[None] | None] = (
             spider_closed_callback
         )
         self.start_time: float | None = None
+        self._start: AsyncIterator[Any] | None = None
+        self._started_request_processing = False
+        downloader_cls: type[Downloader] = load_object(self.settings["DOWNLOADER"])
+        try:
+            self.scheduler_cls: type[BaseScheduler] = self._get_scheduler_class(
+                crawler.settings
+            )
+            self.downloader: Downloader = downloader_cls(crawler)
+            self.scraper: Scraper = Scraper(crawler)
+        except Exception:
+            self.close()
+            raise
```

Comment:
To exit cleanly in case of unhandled exceptions, since this is user-defined code.